pkg/types/azure/disk.go

```diff
@@ -44,5 +46,24 @@ type DiskEncryptionSet struct {
 	Name string `json:"name"`
 }
 
+// VMDiskSecurityProfile specifies the security profile settings for the managed disk.
+// It can be set only for Confidential VMs.
+type VMDiskSecurityProfile struct {
+	// DiskEncryptionSet specifies the customer managed disk encryption set resource id for the
+	// managed disk that is used for Customer Managed Key encrypted ConfidentialVM OS Disk and
+	// VMGuestState blob.
+	// +optional
+	DiskEncryptionSet *DiskEncryptionSet `json:"diskEncryptionSet,omitempty"`
+	// SecurityEncryptionType specifies the encryption type of the managed disk.
+	// It is set to DiskWithVMGuestState to encrypt the managed disk along with the VMGuestState
+	// blob, and to VMGuestStateOnly to encrypt the VMGuestState blob only.
+	// When set to VMGuestStateOnly, the VTpmEnabled should be set to true.
+	// When set to DiskWithVMGuestState, both SecureBootEnabled and VTpmEnabled should be set to true.
+	// It can be set only for Confidential VMs.
+	// +kubebuilder:validation:Enum=VMGuestStateOnly;DiskWithVMGuestState
+	// +optional
+	SecurityEncryptionType *string `json:"securityEncryptionType,omitempty"`
```

Comment:
There are a few different `*string` enums here, but the user input is not being checked. So if a user fat fingers the value or enters an arbitrary string, those issues will potentially not be caught until we have an infrastructure failure.

It would be better to type these values and validate the input. 

Like for example, how we type the Azure Cloud Environment:

https://github.com/openshift/installer/blob/4f59664588c4472b7aba2838159651e729908dff/pkg/types/azure/platform.go#L106-L123

and then check the values:

https://github.com/openshift/installer/blob/4f59664588c4472b7aba2838159651e729908dff/pkg/types/azure/validation/platform.go#L81-L83

You don't necessarily have to type them, especially if that requires too much refactoring, but at a minimum we should be checking that user's can't just enter arbitrary strings that will cause issues once the install is under way.

Reply:
Thanks for this feedback, I absolutely agree that it makes sense to warn the user about potential errors as early as possible. I have included all the necessary validations [here](https://github.com/openshift/installer/pull/7312/files#diff-0d474b476aba944b39d52230861aa5f9b92383c696a81ece3ceabc7dd260b3cfR116-R204). Does this make sense? Or we should do the minimum validations possible, e.g. just validate for the values of the `string` fields? 

PS. these validations are the ones we have in the respective  [MachineAPI provider side](https://github.com/openshift/machine-api-provider-azure/blob/main/pkg/cloud/azure/services/virtualmachines/virtualmachines.go#L454-L543)

Reply:
These validations make sense and thanks for including them. As I learn more about these features, and see your other comment that these features depend on the VM instance type, we should actually include further validations to ensure that users have selected an appropriate instance type. 

I went ahead and tagged this PR, so we can get it merged. But I hope you can add an additional PR to check that the users are selecting a suitable instance type. 

We have existing code that will grab VM capabilities from the Azure SDK and run validations: https://github.com/openshift/installer/blob/0367c39db5f535fffd1d3f95d6cd7aad401d6af5/pkg/asset/installconfig/azure/validation.go#L249-L255
Hopefully this functionality will be exposed through the `capabilities` mechanism and it would be relatively straightforward to extend this validation to ensure that users are selecting appropriate VM types.

Reply:
@patrickdillon thank you for the detailed response, for completeness I'm referencing here the respective [PR](https://github.com/openshift/installer/pull/7469).